Afform - Add support for saving and restoring draft submissions

colemanw opened the pull request:
Overview
----------------------------------------
Adds autosave-draft functionality to FormBuilder submission forms.

Before
----------------------------------------
Afform lacks save-as-draft feature

After
----------------------------------------
When the "Save Draft" form setting is enabled, authenticated users will have their work auto-saved every 10 seconds.
Re-visiting the form will automatically restore their work-in-progress.

[dev/core#5697](https://lab.civicrm.org/dev/core/-/issues/5697)

colemanw committed:
Afform - Add support for saving and restoring draft submissions

Fixes dev/core#5697

When the "Autosave Draft" form setting is enabled, authenticated users
will have their work auto-saved every 10 seconds.

One or more "Save Draft" buttons can also be added to the form for manual control.

Re-visiting the form will automatically restore any saved draft.

danielstrum commented:
Hi @colemanw, Thanks for your quick work again. I did actually want a specific "Save as Draft" button with a very clear notification that the draft has been saved. The forms I am going to be using this for are very long and complex and the user must feel very confident that their work has been saved. I also think they need to have it save at a point of their choice, not automatically.

I could imagine the "save as draft" form settings to have both options, "Auto-save" and "Save a draft button" and the user could chose either one or both? Does that seem like an option?

Also, the form I tested this on is using conditionals based on hidden fields. In this example, the form is for a grant applicant to submit information and documents specific to the way the funds should be sent (bank info, tax status documents...). Earlier in the process they had specified if they were using a fiscal sponsor or not and this had been set in a custom field. The info we need depends on this and the form conditionally uses this field (hidden) to request the proper information. 

What happened with the form is that once this value was set and the form was loaded, it then became "locked" in that condition. Going into the backend and changing the value for that field (from "using a fiscal sponsor" to "not using a fiscal sponsor") did not change the condition of the form. The only way I was able to change it was to completely delete the (test) grant and create a new one (which would not be possible for a real grant with all info already inputted).

Again, thanks for the quick turn around!

colemanw commented:
@danielstrum ok I've revised this PR and added the "Save Draft" buttons. Now:

- You can enable Autosave
- You can drag one or more "Save Draft" buttons onto the form, or
- You can do both

I added user-feedback to the buttons so that its obvious when a draft has been saved. It's pretty neat, check it out.

colemanw commented:
> @danielstrum Also, the form I tested this on is using conditionals based on hidden fields...

Ooh, that's quite tricky. It sounds like you want to exclude certain fields from the "Save Draft" process and have them re-load from the database instead of from the saved draft. Hmm... that's a tough one. I mean, the whole concept of a draft is that every field on the form gets saved as-is so that when the user returns to the form it's exactly like it was before. It sounds like you _don't_ want it to be exactly like it was before... which kinda goes against the idea of a saved draft... 🤔

Could you say a little more about the use-case for having this custom field be updated in that small window of time between when the form is started and when it's submitted?

danielstrum commented:
@colemanw Again... thank for the super quick! I will test the new button functionality and reply. 

As for the other issue with the hidden fields, I think this will not be an issue now that the form will not be saving automatically. By the time a user get to the point of clicking "save as draft", all the hidden fields will have the correct values and will not need to be changed. Please disregard that issue :-)

colemanw commented:
@danielstrum that's great. Glad it worked out. FYI I just made a change to the function name for consistency and force-pushed this PR. If you've already applied the patch, you should delete any "Save Draft" buttons from your form, remove the old patch, and then apply the new one. Sorry for the inconvenience but I get a little obsessive about making the code as clean & consistent as possible.

danielstrum commented:
Hi @colemanw. OK... this seems to be working well and will meet my needs. Again, super thanks!

colemanw commented:
You're welcome @danielstrum - glad it's working well for you!
